### PR TITLE
Added support for better IAM Policy evaluation and basic terraform resource relationship querying

### PR DESCRIPTION
**Pull request by rams3sh**
Closes #

https://github.com/Checkmarx/kics/issues/4739

**Proposed Changes**
-  Added better handling of IAM Policy objects through regex based normalization
-  Added readily usable IAM based functions
-  Added has_relations function to query resource relationship in terraform
- Added privilege escalation actions and managed policies in common.json to be referred across queries
- Added a new query to check lambda functions with privilege roles. 

I submit this contribution under the Apache-2.0 license.

### CODE REVIEW

File: assets/libraries/common.rego [hunk 1]
```diff
@@ -433,3 +433,242 @@ any_principal(statement) {
...
+    not has_atleast_one_match(resource, object.get(statement, "not_resource", []))
+} else {
+	object.get(statement, "resource", []) != []
+    has_atleast_one_match(resource, object.get(statement, "resource", []))
+}
+
+# This method helps in evaluating if a principal matches in a statement.
+# In case true is returned , it just means that this principal's access is affected by this statement. 
+# One has to explicitly check whether it is allowed / denied.
+statement_matches_principal(statement, principal, principal_type) {
```
Comment (rafaela-soares) [inline assets/libraries/common.rego:554]: I think Principal can be also set as a string

**Review by rafaela-soares — CHANGES_REQUESTED** [review]
Such a great contribution 🚀 Thank you!

I have only a few comments/suggestions

File: assets/libraries/common.rego (hunk 1 above)
Comment (rams3sh) [reply assets/libraries/common.rego:554]: Hey yes you are right. But it has been handled in `make_regex_compatible_policy_statement` method. So all statements that are consumed by the functions implemented in this PR requires the policy to be normalized by the `make_regex_compatible_policy_statement`. The values irrespective of being a list or string for Principal / Resource / Actions are normalized to a list by default for easy evaluation later.  You can check that in line 475.

Let me know if any changes are to be done here.

File: assets/queries/terraform/aws/lambda_function_with_privileged_role/query.rego [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+CxPolicy[result] {
+	# For Pre-existing Managed Policy Attachment (i.e not defined within the current terraform template and hard coded as just policy arn)
+	document = input.document[i]
+	lambda = document.resource.aws_lambda_function[function_id]
+	# Checking for role whose id matches in the role of lambda arn reference
+	role = document.resource.aws_iam_role[role_id]
+	terraform_lib.has_relation(role_id, "aws_iam_role", lambda, "role")
+	attached_aws_managed_policy_arns := terraform_lib.get_attached_managed_policy_ids(role_id, "role", input)
```
[the remark this answers is not in the document]
Comment (rams3sh) [reply assets/queries/terraform/aws/lambda_function_with_privileged_role/query.rego:73]: Sure , I will make the required change. Missed it.

File: assets/queries/terraform/aws/lambda_function_with_privileged_role/query.rego [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+# It does not evaluate wholistically across all attached policies.
+# It considers only presence of certain set of actions and it's allowance 
+# in the policy without the context of which resource(s) it applies to.
+
+CxPolicy[result] {
+	# For Inline Policy attachment
+	document = input.document[i]
+	lambda = document.resource.aws_lambda_function[function_id]
+	# Checking for role whose id matches in the role of lambda arn reference
+	role = document.resource.aws_iam_role[role_id]
```
[the remark this answers is not in the document]
Comment (rams3sh) [reply assets/queries/terraform/aws/lambda_function_with_privileged_role/query.rego:22]: Yeaah , you are rite. Changing `input.document.resource` to `input.document[_].resource` should satisfy that case. I will correct it and push it in a while.

File: assets/queries/terraform/aws/lambda_function_with_privileged_role/query.rego (hunk 2 above)
[the remark this answers is not in the document]
Comment (rams3sh) [reply assets/queries/terraform/aws/lambda_function_with_privileged_role/query.rego:73]: Rectified. Please review and let me know now.

File: assets/queries/terraform/aws/lambda_function_with_privileged_role/query.rego (hunk 3 above)
[the remark this answers is not in the document]
Comment (rams3sh) [reply assets/queries/terraform/aws/lambda_function_with_privileged_role/query.rego]: Rectified. Please review and let me know

[the remark this answers is not in the document]
Comment (rafaela-soares) [reply assets/queries/terraform/aws/lambda_function_with_privileged_role/query.rego]: Thank you! It's great!

File: assets/libraries/common.rego (hunk 1 above)
Comment (rafaela-soares) [reply assets/libraries/common.rego:554]: You are right! My bad, sorry!

Can you fix the merge conflicts, please?

**Comment by rams3sh** [comment]
@rafaela-soares  I have resolved the conflcts. Let me know if this looks good now.

**Review by rafaela-soares — APPROVED** [review]
LGTM 🚀

**Comment by rafaela-soares** [comment]
> @rafaela-soares I have resolved the conflcts. Let me know if this looks good now.

Looks amazing! Thank you so much 🚀

**Pull request merged**